services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java

```diff
@@ -252,8 +252,9 @@ private String getIDTokenForLogout(KeycloakSession session, UserSessionModel use
     }
 
     protected void processAccessTokenResponse(BrokeredIdentityContext context, AccessTokenResponse response) {
-
-
+        // Don't verify audience on accessToken as it may not be there. It was verified on IDToken already
+        JsonWebToken access = validateToken(response.getToken(), true);
+        context.getContextData().put(KeycloakOIDCIdentityProvider.VALIDATED_ACCESS_TOKEN, access);
```

Comment:
@rm3l The access token in OIDC (it derives from OAuth2) is not format specified in contrast to ID Token which should be a JWT. So an OIDC complaint broker can return Access Tokens that are not a JWT but opaque  (see for example [this example in the spec itself](https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse) which shows an opaque access token and a JWT ID token).

Therefore the mapper only applies to the ID token in the `OIDCIdentityProvider` but it applies to both in the `KeycloakOIDCIdentityProvider` (as keycloak also uses a JWT access token). We cannot assume that any OIDC providers are going to issue JWT access tokens as you are doing here.

Maybe adding an option in the OIDC provider to mark the access token is a JWT is the solution (false by default), but first we need to know if this is really going to be used by people. What OIDC broker are you using? Do you know many implementations out there using JWT tokens as access tokens as keycloak?

Reply:
@rmartinc Azure AD is one such IDP that sends the access token as a JWT : https://learn.microsoft.com/en-us/azure/active-directory/develop/access-token-claims-reference so this affects any Keycloak users hoping to broker from Azure AD.